File: rlwrap-replace-escaped-lambdas.pl
#!/usr/bin/env perl

use lib ($ENV{RLWRAP_FILTERDIR} or ".");
use RlwrapFilter;
use strict;

my $filter = new RlwrapFilter;

$filter -> input_handler(\&expand_symbols);
$filter -> history_handler(\&expand_symbols);
$filter -> completion_handler(\&complete_symbol);
$filter -> run;

sub expand_symbols {
    my ($i) = @_;

    my $lambda_re = qr/\\l(ambda|ambd|amb|am|a|)/;
    my $Lambda_re = qr/\\L(ambda|ambd|amb|am|a|)/;
    my $bottom_re = qr/\\b(ottom|otto|ott|ot|o|)/;
    my $forall_re = qr/\\f(orall|oral|ora|or|o|)/;

    $i =~ s/$lambda_re/λ/;
    $i =~ s/$Lambda_re/Λ/;
    $i =~ s/$bottom_re/⊥/;
    $i =~ s/$forall_re/∀/;

    return $i;
}

sub complete_symbol {
    my($line, $prefix, @completions) = @_;

    my $lambda_re = qr/\\l(ambda|ambd|amb|am|a|)/;
    my $Lambda_re = qr/\\L(ambda|ambd|amb|am|a|)/;
    my $bottom_re = qr/\\b(ottom|otto|ott|ot|o|)/;
    my $forall_re = qr/\\f(orall|oral|ora|or|o|)/;

    if ($prefix =~ /$lambda_re$/) {
        $prefix =~ s/$lambda_re$/λ/;
        unshift @completions, $prefix;
    }

    if ($prefix =~ /$Lambda_re$/) {
        $prefix =~ s/$Lambda_re$/Λ/;
        unshift @completions, $prefix;
    }

    if ($prefix =~ /$bottom_re$/) {
        $prefix =~ s/$bottom_re$/⊥/;
        unshift @completions, $prefix;
    }

    if ($prefix =~ /$forall_re$/) {
        $prefix =~ s/$forall_re$/∀/;
        unshift @completions, $prefix;
    }

    return @completions;
}

File: rlwrap-replace-escaped-lambdas.py
#!/usr/bin/env python3

import sys
import os

if 'RLWRAP_FILTERDIR' in os.environ:
    sys.path.append(os.environ['RLWRAP_FILTERDIR'])
else:
    sys.path.append('.')

import re

lambda_re = "\\\\l(ambda|ambd|amb|am|a|)"
Lambda_re = "\\\\L(ambda|ambd|amb|am|a|)"
bottom_re = "\\\\b(ottom|otto|ott|ot|o|)"
forall_re = "\\\\f(orall|oral|ora|or|o|)"

lambda_end_re = lambda_re + "$"
Lambda_end_re = Lambda_re + "$"
bottom_end_re = bottom_re + "$"
forall_end_re = forall_re + "$"

def expand_symbols(i):
    i = re.sub(lambda_re,"λ",i)
    i = re.sub(Lambda_re,"Λ",i)
    i = re.sub(bottom_re,"⊥",i)
    i = re.sub(forall_re,"∀",i)
    return i

def complete_symbol(line, prefix, completions):
    if re.search(lambda_end_re, prefix):
        return [re.sub(lambda_end_re, 'λ', prefix)] + completions
    if re.search(Lambda_end_re, prefix):
        return [re.sub(Lambda_end_re, "Λ", prefix)] + completions
    if re.search(bottom_end_re, prefix):
        return [re.sub(bottom_end_re, "⊥", prefix)] + completions
    if re.search(forall_end_re, prefix):
        return [re.sub(forall_end_re, "∀", prefix)] + completions

    return completions

import rlwrapfilter

filter = rlwrapfilter.RlwrapFilter()
filter.input_handler = expand_symbols
filter.completion_handler = complete_symbol
filter.run()
